antsibull/schemas/base.py

```diff
@@ -152,7 +152,7 @@ def handle_renamed_attribute(cls, values):
 REQUIRED_COLLECTION_NAME_OR_EMPTY_STR_F = p.Field('', regex='^([^.]+\\.[^.]+)?$')
 
 #: Constrained string listing the possible types of a return field
-RETURN_TYPE_F = p.Field('str', regex='^(bool|complex|dict|float|int|list|str)$')
+RETURN_TYPE_F = p.Field('str', regex='^(bool|complex|dict|float|int|list|str|path|raw)$')
```

Comment:
I don't think we'd want raw in the output.  Maybe that's something that should be changed in ansible-core.  path, I guess would be okay although if it's not being used in anything right now, I wonder if it's not a useful distinction from str and should be removed from core as well.

Reply:
I think distinguishing between `str` and `path` is important from a docs point of view. As a user, it makes it easier to see whether something returned is a path, or just a 'random' string.

About `raw`: this basically is a compatibility option for plugins which return different things (depending on input, or 'random' other circumstances). Since `type` should always be specified, one needs to put something in there. Putting `string or dict or int` or something like that will cause validation errors to show up on the docs page, which is just horrible, see for example https://ansible.fontein.de/collections/community/general/xfconf_module.html#authors Having something like that in the documentation of a module/plugin I care about is the worst thing that can happen IMO. `raw` is the only way to avoid this to show up (well, next to just not documenting the return value, which is also not really acceptable).

Reply:
You've sold me on the value of path: +1 to that.

We talked about raw on IRC a bit and I'm not sure what to do about that yet. Some things that came out of the discussion:
* playbooks provide a bad environment for programming so something that can return any type feels like it is a design wart in the module.  It would be better for the module author to use different fields for each type. 
* Module authors won't always consider that, though, and omitting the return docs for that item because the module author made a poor design choice is also not a good solution.
* In the xfconf example, the module should have been coded to always return a `list of str` instead of trying to return `str or list of str`
* raw in the arg_spec is slightly different than what we're doing here.  raw in the argspec means "ansible will not convert this value.  The module should be documenting what you can pass in".  raw in here means "The module can return more than one type of value.  If you want to use the return programatically, you'll need to check what the return type is in your cases and write playbook code to handle all of those types".
  * There's a much higher burden on the playbook author here (which, unfortunately, only the module author can fix)
  * We're not specifying how ansible will treat this value; instead we're specifying what hoops the playbook author has to deal with.

So I think for me personally, there's two things I don't like about raw in return values:

1. It's helping to code bad return type design.
2. The purpose of the return type feels different enough from its arg_spec usage that I don't like reusing the name `raw`.  `any`, `anytype`, `unknowntype`, `multipletypes`, `multitype` all seem better here.

I think I can resign myself to 1.  If we feel strongly enough about that, we should encode that as a collection criteria, otherwise we need to document what we're given.

Can we rename it for 2., though?  IOr is raw already being used to mean this on the core side?

Reply:
It's already used (partially by me), see for example https://github.com/ansible-collections/community.general/pull/881/files#diff-daf11feb54ca616c9c06f47d97eace504658c95e9b2e408ac935b6f22c02a922R79 and https://github.com/ansible/ansible/pull/71736/files#diff-90bb7bcd81518d5f4e00409a2ed162949fb24fee3bf07aca7d77cab0c5632ae3R56 (in some cases I used it because it was hard from the code to see what exactly it returns, and in some cases the code returned different types)

Reply:
felix and I agreed on IRC to use `any` for this but to accept `raw` and convert it into `any`.  `any` conveys the meaning that the return value can be of any type.  `raw` would convey the meaning of unprocessed which is not what we mean.

I approve this PR with that change made.